### PR TITLE
Adding interface, implementation, and tests for vulnerability capabilities for a package domain model

### PR DESCRIPTION
**Pull request by jebriede**
# Bug

Fixes:

## Description
Encapsulates the vulnerability list and logic. Provides an interface for use in a package model type as well as the implementation and unit tests.

## PR Checklist

- [ ] Meaningful title, helpful description and a linked NuGet/Home issue
- [x] Added tests
- [ ] ~~Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.~~

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        private IEnumerable<PackageVulnerabilityMetadataContextInfo> _vulnerabilities;
+
+        public IEnumerable<PackageVulnerabilityMetadataContextInfo> Vulnerabilities
+        {
+            get => _vulnerabilities;
+            private set => _vulnerabilities = value.OrderByDescending(v => v?.Severity ?? -1);
+        }
+
+        public bool IsVulnerable => Vulnerabilities.Any(v => v != null);
```
Comment (jgonz120) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: so you're saying it's possible to have a list of vulnerabilities where all entries are null?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Models/Package/VulnerableCapabilityTests.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+            // Assert
+            Assert.Equal(expected, result);
+        }
+
+        [Fact]
+        public void IsVulnerable_NullVulnerabilityInCollection_ReturnsFalse()
+        {
+            // Arrange
+            PackageVulnerabilityMetadataContextInfo[] vulnerabilities = [null];
```
Comment (jgonz120) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Models/Package/VulnerableCapabilityTests.cs]: is this currently possible? Should we consider blocking this all together?

**Review by jgonz120 — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 1 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: It is, programmatically. It doesn't make much sense to have that, but there a few strategies I can think of:
1) Prevent the construction of a `VulnerableCapability` if the collection of vulnerabilities contains any nulls. This would require iterating over the collection upon construction to validate if any of the entries are null. With big lists this could take a performance hit for constructing many Packages with IVulnerable and I would not favor this.
2) Allow the collection to contain nulls because it's just a list of objects, but handle them gracefully so we essentially treat them as if they did not exist. This is what I've done in this PR and made that behavior explicit by adding a test around it and to ensure that it is handled gracefully.

Can you think of another way you'd suggest instead?

Comment (jgonz120) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: Yea, I think iterating through them to validate there aren't any nulls seems excessive. I guess I didn't realize we had been accounting for this in the old code too.

Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: @jgonz120 We aren't accounting for it explicitly in the ViewModels today which presents a potential flaw. The goal here is to make this code more robust than what we had before and put it under test. By handling this edge case gracefully, we avoid any NRE's in case a null makes its way into the list at any point.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+        public PackageVulnerabilitySeverity VulnerabilityMaxSeverity
+        {
+            get
+            {
+                // Vulnerabilities are ordered so the first element is always the highest severity
```
Comment (martinrrm) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: This is great for performance, not sure if we are currently iterating. More context: https://learn.microsoft.com/en-us/nuget/api/vulnerability-info#vulnerability-page.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: If this is not officially part of the contract, we shouldn't depend on it.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/IVulnerable.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+using System.Collections.Generic;
+using NuGet.Protocol;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.UI
+{
+    interface IVulnerable
+    {
+        public IEnumerable<PackageVulnerabilityMetadataContextInfo> Vulnerabilities { get; }
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/IVulnerable.cs]: Can you use more concrete types to avoid paying the enumerator costs?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 1 above)
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: Make this class immutable and avoid paid the cost of `Any`.
In general, I'd avoid properties that force a compute that's not cached.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 3 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: Correct, there's no way to guarantee the list will be ordered from any and all vulnerability sources, both current and future. This comment points out that the list is ordered. It's ordered in this class (see the setter) explicitly to avoid making assumptions about the data source. Does that help? Do you have any suggested edits?

Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: It does, right? `The list of known vulnerabilities for a package should be sorted in descending order of the max version of the version range`.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: Doesn't that talk about the way the version ranges are sorted and not the list of vulnerabilities?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 1 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: This file has `#nullable enable`, and the definition of the `Vulnerabilities` property doesn't allow nulls, so either the definition is wrong, or this not null check shouldn't be needed.

Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: Additionally, the current implementation of calculating the `IsVulnerable` property gets the `Vulnerabilities` property, which in turn uses LINQ's `OrderByDescending` API.  Therefore, in order to determine if the current package is vulnerable, the code is going to allocate a list of equal or greater capacity to the `_vulnerabilities` enumerable, spend CPU cycles sorting it, then just check if there's a first item, without caring about the contents of it. The `.Any()` call causes an enumerator to be allocated on the stack, as well as the `OrderByDescending` call. Both of these methods pass in a predicate, and unless the C# compiler has improved since I last checked, they both cause heap allocations as well (could be avoided by creating class methods and passing it without using an anonymous function, but that makes the code less pretty).

Referring to Nikolche's other comment about using a concrete class, all these heap allocations could be avoided by using a concrete class, and making this property's body `=> _vulnerabilities.Count > 0`

Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: > Additionally, the current implementation of calculating the `IsVulnerable` property gets the `Vulnerabilities` property, which in turn uses LINQ's `OrderByDescending` API. Therefore, in order to determine if the current package is vulnerable, the code is going to allocate a list of equal or greater capacity to the `_vulnerabilities` enumerable, spend CPU cycles sorting it, then just check if there's a first item, without caring about the contents of it.

> The .Any() call causes an enumerator to be allocated on the stack, **as well as the OrderByDescending call**

This is not correct. The OrderByDescending is in the setter of the Vulnerabilities property, not in the getter. Therefore, it will only be sorted once when assigned privately, not upon each access.

Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: > This file has `#nullable enable`, and the definition of the `Vulnerabilities` property doesn't allow nulls, so either the definition is wrong, or this not null check shouldn't be needed.

I see what you mean. Both the tests and the `VulnerabilityCapability` classes themselves have #nullable enable yet the Test allows a null value to be used. As such, it proves that with nullable enabled on the caller, we can still end up with nulls making their way into the list. While a test can show that it's possible, we should have code that handles this edge case and a test that proves it's being handled. If I wrote the test incorrectly though, please let me know how and I'll fix it!

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/IVulnerable.cs (hunk 4 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/IVulnerable.cs]: After discussing offline, we agreed on using `IReadOnlyList` to preserve immutability, have improved Count performance, and provide the ability to use `NoAllocEnumerate` if used on a hot code path.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 1 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: @zivkan I removed the null checks and the test. This bug is open against Roslyn https://github.com/dotnet/roslyn/issues/72142 which is erroneously allowing the list to have a null in only one case that I could tell. I'm deciding not to code around this bug for an unlikely edge case. Let me know what you think!

**Review by jgonz120 — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace NuGet.PackageManagement.UI
+{
+    public class VulnerableCapability : IVulnerable
+    {
+        private IReadOnlyList<PackageVulnerabilityMetadataContextInfo> _vulnerabilities = [];
+        public IReadOnlyList<PackageVulnerabilityMetadataContextInfo> Vulnerabilities
+        {
+            get => _vulnerabilities;
+            private set => _vulnerabilities = [.. value.OrderByDescending(v => v.Severity)];
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: I think I already made a similar comment in another thread, but I don't think that sorting this list in a way that requires a bunch of heap allocations (especially since it will create a full duplicate list, it's not sorting in-place) is worth the cost to have a "nice" internal API. 

I imagine the CPU cost of sorting will be higher than having `VulnerabilityMaxSeverity` use a `for` loop, even if the sort was zero memory allocation, when there's more than one or two items in the collection.  This syntax hides the fact that it has to do a `.ToList()` in addition, so even if there's a single item in the collection, allocating the same array 3 times (once for the list passed to the setter, once in `OrderByDescending`, and finally the array that gets assigned to the field) means that this is much more costly than a for loop that gets called a small number of times.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            get
+            {
+                if (!IsVulnerable)
+                {
+                    throw new InvalidOperationException("Vulnerabilities is empty");
+                }
+
+                // Vulnerabilities are ordered on set so the first element is always the highest severity
+                int severity = Vulnerabilities[0].Severity;
+                if (Enum.IsDefined(typeof(PackageVulnerabilitySeverity), severity))
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs:41]: I know that some `Enum` class static methods cause heap allocations on .NET Framework, in addition to bing relatively slow, which is why `AuditUtility.cs` doesn't use `Enum.ToString()` or `Enum.Parse`.  I assume that `Enum.IsDefined` has the same issue.

https://github.com/NuGet/NuGet.Client/pull/5361#discussion_r1295613063

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 1 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: This is @jgonz120's thread 😁

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 5 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs:20]: The list needs to be ordered for usage other than simply for the MaxSeverity. It will always need to be ordered for display, which is the main purpose for the vulnerability data existing in this list.

There are some assertions you've made that are not accurate. I'll address them each:

> sorting this list in a way that requires a bunch of heap allocations

"A bunch of heap allocations" is a subjective statement. Specifically, the cost will be 2x the size of the list. Even when considering this for multiple packages, the cost is negligible, especially because most packages will have an empty list. The majority of packages in a solution have 0 vulnerabilities, and the ones that do are typically small lists, as you've acknowledged.

> so even if there's a single item in the collection, allocating the same array 3 times (once for the list passed to the setter, once in OrderByDescending, and finally the array that gets assigned to the field)

This statement is attributing the allocation of a list that is passed into the constructor to this method of sorting. That cost will be incurred no matter what and cannot be attributed to the sorting or this class. Therefore, to say it will cost 3x the size of the list to sort in this manner is a misrepresentation of what the cost is.

> means that this is much more costly than a for loop that gets called a small number of times.

This is subjective. On one hand we allocate memory for a sort we need to do anyway. We can delay it to the View or ViewModel code. On the other hand we spend more CPU cycles and have to sort at some point anyway.

Overall, I thank you for taking the time to point out these issues. I've made a conscious decision on the tradeoffs and view this as premature optimization. I will leave it as is and consider this resolved.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 6 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs:41]: Thank you for pointing out a potential allocation. You say you assume that Enum.IsDefined has the same issue as Enum.ToString and Parse. Have you verified this? I'd be interested in learning more about it.

What do you suggest instead? We are making a conversion between an int with the existing contract class to an Enum. We want to make sure that we handle said Enum correctly in case a value that is not correct is supplied.

I look forward to your actionable, alternative suggestions.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 5 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: If the data needs to be ordered for other reasons, for example for display, this information isn't obvious. There's no comment either on this class, nor the interface, that sorted is part of the contract. The BCL has a `SortedList` type, which isn't being used here (but to be fair, it's much more a sorted dictionary than a list). There's also no mention about it in the PR description (but that wouldn't be a good place for an interface contract anyway). While there is a test that ensures it's sorted, it's not clear if this is testing the implementation, or if it's supposed to be part of the contract.

But if it needs to be sorted, a way to do which reduces the number of duplicated array allocations is to duplicate into a `List<T>` and then call [`List<T>.Sort`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.sort?view=net-9.0#system-collections-generic-list-1-sort). It sorts in-place, and as long as the key comparer doesn't cause allocations, sorting can be zero allocations.

Side note, I had a look at Reference Source, and it appears that it creates [one](https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,2598), [two](https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,2649) arrays, so it's worse than I originally thought. I wonder why they didn't implement is as `OrderBy(...) { var list = new List<T>(enumerable); list.Sort(); return list; }`.

Also, a different design choice is to make it the responsibility of the code that creates instances of `VulnerabilityCapability`, so whatever code discovered the relevant advisories for the current package, it can use `List<T>.Sort` and no duplicate collection is required. If it provides this class with a sorted list, this class can hold a reference without making a copy. This won't give us compiler enforcement of the contract.

It's really a shame that there is no `SortedList<T>` that can be cast to `IReadOnlyList<T>`, because then we would have been able to have a class constructor take in the sorted list and have the vulnerability list be a get only read only list property.

If compiler enforced contracts are important enough to you, it's possible to create our own `SortedList<T>` that wraps `SortedList<T, T>`, and makes the indexer and enumerator return the `T`, rather than `KeyValuePair<TKey, TValue>`.

My opinion is that since this class won't ship in a package, and therefore nobody outside the NuGet.Client can call it, compiler enforced contractual correctness is not as important as designing with performance as one of its multiple features.

Most of all, I believe we can create abstractions that meet both our reducing complexity requirement, as well as performance requirement.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 6 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs:41]: When is it the PR author's responsibility to do investigations, and when is it the reviewer's responsibility? Honest question because there are times I create my own PRs and someone gives me feedback I don't agree with. I'd like to know when I should spend the time, or ask the reviewer to provide me with more information.

[The API being used is `IsDefined(Type, Object)`](https://learn.microsoft.com/en-us/dotnet/api/system.enum.isdefined?view=net-9.0#system-enum-isdefined(system-type-system-object)), and casting an `int` to `Object` is the very definition of boxing. 

Indeed, [using sharplab to look at the IL](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs4fuAglzoBZABQBKNhy5TCAdjoBRAHYBXALYA6AJIBnACIwAZtkUwAJsIwBPAA4wIB4YVLqAcjAzqAEhgzWAyhgAhhjK2kwQpjCiaHSkxDSiANySnAC+VKlAA=) shows:

> IL_000c: box [System.Runtime]System.Int32

The PR that I linked added the comment "Enum parsing and ToString are a magnitude of times slower than a naive implementation." to 4 methods, and picking the first one, we can see the method implementation:

https://github.com/NuGet/NuGet.Client/blob/9202ddad5fabd4d7737fa0c717524fbe2455c972/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/AuditUtility.cs#L516-L539

Also looking at the other 3 methods with the same comment, the pattern is that we write our own code that enumerates all the values. So here's one idea:

```cs
int MethodThatNeedsABetterName(int value)
{
     switch ((PackageVulnerabilitySeveiry)value)
     {
        PackageVulnerabilitySeverity.Low:
        PackageVulnerabilitySeverity.Moderate:
        PackageVulnerabilitySeverity.High:
        PackageVulnerabilitySeverity.Critical:
            return value;

       default:
            return PackageVulnerabilitySeverity.Unknown;
     }
}
```

Another idea that results in less code is `int Method(int value) => (value >= PackageVulnerabilitySeverity.Low && value <= PackageVulnerabilitySeverity.Critical) ? value : PackageVulnerabilitySeverity.Unknown;`

Since this is asynchronous communication and I can't predict if you'll respond that this has a risk of breaking if `PackageVulnerabilitySeverity` changes, I'll address that potential concern, just in case:

NuGet has copied this seveiry definition from Github Advisories.  [Looking at one advisory](https://github.com/advisories/GHSA-gjf6-3w4p-7xfh), under the "Severity" heading, the score has a dashed underline, turns out to be clickable, which pops up a tooltip saying it's based on the Common Vulnerability Scoring System (CVSS). Doing a search, it appears that [it's defined here](https://www.first.org/cvss/). The CVSS V4 spec has a heading "Qualitative Severity Rating Scale" that defines low, medium, high, and critical labels for the various score ranges. The same is true for [CVSS V3.1](https://www.first.org/cvss/v3.1/specification-document) and [CVSS V3.0](https://www.first.org/cvss/v3.0/specification-document). CVSS 2 and CVSS 1 do not appear to have any equivalent. In other words, this qualitative severity has never changed since it was first created, which wikipedia says CVSS 3.0 came out in 2015.

While it's possible that CVSS v5 or later could change it, I expect the people who define CVSS are aware this would impact a lot of security software, and similarly GitHub would need to choose to migrate off CVSS 4.0 to whatever version hypothetically breaks the severity labels.  All this to say that I'm confident enough that this won't change, in order to have taken the perf benefit of zero allocations for ToString and Parse in AuditChecker.cs.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        public VulnerableCapability(List<PackageVulnerabilityMetadataContextInfo> vulnerabilities)
+        {
+            if (vulnerabilities == null)
+            {
+                throw new ArgumentNullException(nameof(vulnerabilities));
+            }
+
+            Vulnerabilities = vulnerabilities;
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs:59]: This can be shortened, if you prefer

```suggestion
            Vulnerabilities = vulnerabilities ?? throw new ArgumentNullException(nameof(vulnerabilities));
```

**Review by martinrrm — APPROVED** [review]

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Models/Package/VulnerableCapabilityTests.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                ? [new(new Uri("http://example.com"), severity)]
+                : [];
+            var vulnerableCapability = new VulnerableCapability(vulnerabilities);
+
+            // Act
+            var result = vulnerableCapability.IsVulnerable;
+
+            // Assert
+            Assert.Equal(expected, result);
+        }
```
Comment (Nigusu-Allehu) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Models/Package/VulnerableCapabilityTests.cs:33]:
```suggestion
        [Fact]
        public void IsVulnerable_WhenVulnerabilitiesExist_ReturnsTrue()
        {
            // Arrange
            var vulnerabilities = new List<PackageVulnerabilityMetadataContextInfo>
            {
                new PackageVulnerabilityMetadataContextInfo(new Uri("http://example.com"), severity: 1)
            };
            var capability = new VulnerableCapability(vulnerabilities);

            // Act
            bool isVulnerable = capability.IsVulnerable;

            // Assert
            Assert.True(isVulnerable);
        }

        [Fact]
        public void IsVulnerable_WhenNoVulnerabilitiesExist_ReturnsFalse()
        {
            // Arrange
            var capability = new VulnerableCapability(new List<PackageVulnerabilityMetadataContextInfo>());

            // Act
            bool isVulnerable = capability.IsVulnerable;

            // Assert
            Assert.False(isVulnerable);
        }
```

**Review by Nigusu-Allehu — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 6 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs:41]: Thanks @zivkan! We chatted offline and we agreed to defer this for this now, but we'll be aware of opportunity for avoid allocation as the this evolves. Thanks for taking the time to provide this detail!

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs (hunk 5 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/VulnerableCapability.cs]: > a way to do which reduces the number of duplicated array allocations is to duplicate into a List<T> and then call [List<T>.Sort](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.sort?view=net-9.0#system-collections-generic-list-1-sort)

@zivkan thank you for this suggestion! As we chatted about offline, I'll keep note of this optimization for future scenarios like this as well. Sorting a list in place when we know the underlying concrete type is a great idea. I think my latest iteration addresses this now. Please have a look and let me know!

**Review by zivkan — APPROVED** [review]
